### DISCUSSION TITLE
A "FormView" abstraction.

### DISCUSSION THREAD

**Discussion by fadedbee** [gyscos/cursive#773] [Ideas]
At the moment I am manually creating many EditView fields inside of Panels.

I then write functions which populate the contents of the fields from a model struct, and functions which update fields of the model struct from the Panel's EditViews.

I feel that Cursive is missing a FormView abstraction, which would keep EditView fields and a struct in sync (with validation) or a perhaps a more React-structured cyclic binding (rather than direct bi-directional bindings).

Such an abstraction would have a `derive` macro, similar to `serde` so that ViewForms could be automatically generated from annotated structs.

**Comment by gyscos** [comment]
Hi!

Indeed, such a view sounds like it would be very useful! As you mentioned, a React-like structure would be great here. I didn't do this initially because I didn't know of a popular reactive lib (a widespread reactive type would be ideal to integrate with other processing libs). Maybe picking one, or even implementing it here, would be sufficient, without the need for it to be widespread.

**Reply by gyscos** [reply]
I have since started working on the [`feature/rx` branch](https://github.com/gyscos/cursive/tree/feature/rx) which adds `Rx<T>` type for a kind of simplified react-like binding. It's a generalization of the `TextContent` we had for `TextView`, now available for `EditView` and with some extra features.

Behind the scenes it's essentially an `Arc<Mutex<T>>` with some "watchers" (callbacks triggered on a value change). There's some tricks to avoid cloning and locking too much, so recursive cases are possible:

```rust
use cursive_core::utils::Rx;

let rx = Rx::new(String::new());
rx.map(|c| c.len())
  .map(|len| format!("This sentence has {len} chars."))
  .pipe_to(&rx);

assert_eq!(rx.get(), String::from("This sentence has 27 chars."));
```

It should be easier now to derive a view tree with `EditText` from a struct with a bunch of `Rx<String>` values.
